src/main/kotlin/org/rust/ide/annotator/RsFormatMacroAnnotator.kt

```diff
@@ -0,0 +1,127 @@
+            if (range != null && !range.isEmpty())
+                createAnnotation(
+                    ARG_HIGHLIGHTING,
+                    range.toSourceRange(),
+                    null)
+                    .textAttributes = color.textAttributesKey
+        }
+
+        val key = RsColor.VALID_STRING_ESCAPE
+        val highlightSeverity = if (isUnitTestMode) key.testSeverity else HighlightSeverity.INFORMATION
+        for (arg in arguments) {
+            currentOffset = 0
+            val textRange = arg.range.toSourceRange()
+            holder.createAnnotation(highlightSeverity, textRange, null)
+                .textAttributes = key.textAttributesKey
+
+            if (arg.groups[3] != null) {
+                holder.createErrorAnnotation(
+                    arg.range.toSourceRange(),
+                    "Invalid format string: unmatched '}'")
+            }
+            if (arg.groups[1] != null) {
+
+                val inside = arg.groups[2]!!
+                currentOffset = inside.range.first
+                val matchResult = formatArgParser.find(inside.value)!!
+
+                if (!arg.groups[1]!!.value.endsWith("}")) {
+                    val possibleEnd = matchResult.value.length - 1
+                    holder.createErrorAnnotation(
+                        (possibleEnd..possibleEnd).toSourceRange(),
+                        "Invalid format string: } expected.\nIf you intended to print `{` symbol, you can escape it using `{{`")
+                    continue
+                }
+
+                val validParsedEnd = matchResult.range.last + 1
+                if (validParsedEnd != inside.value.length) {
+                    holder.createErrorAnnotation(
+                        (validParsedEnd until inside.value.length).toSourceRange(),
+                        "Invalid format string")
+                } else {
+                    holder.highlightArgument(matchResult.groups["id"]?.range)
+                    holder.highlightArgument(matchResult.groups["width"]?.range)
+                    holder.highlightArgument(matchResult.groups["precision"]?.range)
+                    holder.highlightArgument(matchResult.groups["type"]?.range, RsColor.FUNCTION)
+                    //TODO: argument type/count/other checks
+                }
+            }
+        }
+    }
+
+    private fun macroToFormatPos(macro: String): Int = when (macro) {
+        "println" -> 0
+        "print" -> 0
+        "eprintln" -> 0
+        "eprint" -> 0
+        "format" -> 0
+        "write" -> 1
+        "writeln" -> 1
+        else -> -1
+    }
+
+    companion object {
+        private val ARG_HIGHLIGHTING = HighlightSeverity(
+            "ARG_HIGHLIGHTING",
+            HighlightSeverity.INFORMATION.myVal + 1
+        )
+        val formatParser = Regex("""\{\{|}}|(\{([^}]*)}?)|(})""")
+        const val argument = """([a-zA-Z_][\w+]*|\d+)"""
+        val formatArgParser = Regex("""
+^(?<id>$argument)?
+(:
+(.?[\^<>])?[+\-]?\#?0?
```

Comment:
Redundant character escape? 
![image](https://user-images.githubusercontent.com/3221931/74440820-c3bd5600-4e7f-11ea-8a9b-74751e0f6bd2.png)

Reply:
AFAIK, it is false positive warning, but i will check this again to be sure.

Reply:
Yep, it is false positive when comments are enabled in regexp.